File: balanced_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# TC: O(n)
# SC: O(h)
class Solution:
    def __init__(self):
        self.isValid = True

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root == None:
            return True
        self.height(root)
        return self.isValid

    def height(self, node):
        if node == None:
            return -1
        lh = self.height(node.left)
        rh = self.height(node.right)
        if abs(lh-rh) > 1:
            self.isValid = False   
        return max(lh, rh) + 1

File: balancedBinaryTree.cpp
// Time Complexity : O(n) n is number of nodes in the tree
// Space Complexity : O(n) in case if tree is skewed, O(log n) if the tree is not skewed on one side
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

#include <iostream>
using namespace std;  

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    int __isBalanced(TreeNode* root, bool &flag)
    {
        if(root==nullptr)
        {
            return 0;
        }
        if(flag)
        {
            int leftHeight = __isBalanced(root->left,flag);
            int rightHeight = __isBalanced(root->right,flag);
            int height = abs(leftHeight-rightHeight);
            if(height>1)
            {
                flag = false;
                return -1;
            }
            return 1 + max(leftHeight,rightHeight);
        }
        return -1;
    }
public:
    bool isBalanced(TreeNode* root) {
        bool flag = true;
        __isBalanced(root,flag);
        return flag;
    }
};